flink-metrics/flink-metrics-jmx/src/main/java/org/apache/flink/metrics/jmx/JMXReporter.java

```diff
@@ -541,5 +552,40 @@ public void stop() throws IOException {
 				}
 			}
 		}
+
+		/*
+		 * Better to use the internal API than re-invent the wheel.
+		 */
+		@SuppressWarnings("restriction")
+		private class JmxRegistry extends sun.rmi.registry.RegistryImpl {
```

Comment:
With Java 9+ and the module system, access to such internal classes is harder (gives a strong warning, may be not possible in the future). Could one rewrite this code to wrap an existing registry and delegate all method calls, rather than to extend from this class?

Reply:
Unfortunately not, I haven't found a way not use the internal classes.

Reply:
This is certainly a bit problematic.

We will need to check whether this work as is on Java 11+ or whether we have to do any module export business at runtime (ideally we don't, it may not be worth the effort).

I don't think wrapping the registry is feasible either, the setup is done in the constructor so any delegation we could setup would happen after the registry is already accessible.
Creating a custom RMI registry free of `sun` references doesn't seem realistic either.

From what I can tell we can't get around using internal classes, be it either directly or via reflection.

I've thought about it some more and maybe we should in fact just drop the port option. We are replacing code that we already did not really understand with other code we don't really understand.
This may be something we shouldn't be trying to workaround.